keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,449 @@
+* The API will be read-only and accessible for on-node workloads (we will use a
+  unix socket for the connection) without authorization.
+* The API will be versioned.
+
+### What kind of API to chose?
+
+Our goal is to expose the API locally on the node and without Control Plane
+dependency. That means that we want to avoid authorization which will be done via
+kube-API. To fullfil this requirement we decided to use unix socket to access
+the API. On top of unix socket we can create REST or gRPC API.
+
+|   | gRPC | REST |
+| - | ---- | -----|
+| Support for unix socket | yes | yes |
+| Can reuse existing API | possibly PodResource API. See [here](#can-we-integrate-with-podresource-api) | no, the existing REST API is served on ReadOnlyPort (shutdown on various cloud providers due to security reasons), AuthenticatedPort (authorization is done by kube-API). |
+| Can reuse existing libraries (client-go, helpers in documentation) | no | yes|
+| Support for streaming | yes | yes|
+| Strongly typed | yes | no |
+| Versions support | yes | yes |
+
+Given the above table, we believe that a gRPC API is the best path forward.
+
+### Can we integrate with PodResource API?
+
+|   | PodResource API | Status API |
+| - | --------------- | ------------- |
+| What information is returned with API | Container resources (devices, cpu_ids, memory, dynamic resources) | for Static and Regular Pods pod readiness Conditions and UID  |
+| Who is the owner of the data returned by API | Kubelet, those data are not reported to Control Plane | static Pods - Kubelet, regular Pods - Kubelet, data are reported to kube-API |
+| What is the alternative for the API | reading the state files resource manager use | for regular Pods - fetching pod's data with kube-API, for static Pods - none |
+| Control Plane dependency | no | no |
+| Authorization | none, using unix socket to limit the clients to local workloads | none, using unix socket to limit the clients to local workloads |
+| Access | Read-Only | Read-Only|
+| Clients type / use cases | Device Plugins, CSI, Topology aware scheduling | CNI, monitoring and observability tools |
+
+The PodResource API includes an entirely unrelated set of information that is
+unlikely to be of use to the set of clients that would benefit from
+understanding Pod readiness. We propose creating a new API for this purpose.
+
+### Risks and Mitigations
+
+This API is read-only, which removes a large class of risks.
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests to the API impacting the Kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
```

Comment:
s/malformed/malicious/

This is a little strong to say it *cannot* be used this way, but if SIG security is OK with this I wouldn't argue.

I could imagine if there were some local service that enforced some policy, knowing that it is not yet ready could be useful to an attacker.

(non-blocking comment)

Reply:
done